diff --git a/CMakeLists.txt b/CMakeLists.txt
index 456e86c38..2c5eea1a7 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -60,6 +60,7 @@ enable_testing()
 find_package(Threads REQUIRED)
 include(Doxygen)
 include(GNUInstallDirs)
+include(XalanDebug)
 include(XalanDLL)
 include(XalanWarnings)
 include(XalanFunctions)
diff --git a/KEYS b/KEYS
index 0fc827cde..dac23d8b3 100644
--- a/KEYS
+++ b/KEYS
@@ -164,3 +164,322 @@ JVs/N4xcYH+5vEGou2xoE/X6uonxhbbmtGisQRfFf/kVDMcB7eD2HrJZoQy2AwVp
 3JQv88HXJk79iaam8tc=
 =MBu4
 -----END PGP PUBLIC KEY BLOCK-----
+
+pub   rsa4096 2011-12-30 [SC] [expires: 2021-09-16]
+      F33D281D470AB4436756147C07B3C8BC4083E800
+uid           Roger Leigh <rleigh@apache.org>
+uid           Roger Leigh <rleigh@codelibre.net>
+
+-----BEGIN PGP PUBLIC KEY BLOCK-----
+
+mQINBE790wIBEAC5YfkqgjQIYSHfVvBr5aqjWTbRq4ib8IsEFPyXO69ia7znLyBk
+Sorhyh74DFZIf+OR+4s0mF7XVN9gtJZjmlzM4jomPw8JBAY2/dpcakWOdXp7sVQe
+qNFD93COYF0T/io5kbQDrEBCirooKLHMZt6HefMcJOS5vje4BlPPdBv1R9tb1lNR
+bkV0wDtxlPFGpEoxgjdX5bu99Gk8UB+610cYc+j+OtMOU7uQjCU1JjbPRHTd+qWY
+scYgN68kvdIAWmDAE0ipqJ3AwsLD6F0YAC9kkj/Ie8xEK2tkeH9nifysggecOvJ1
+6jbuZpfGF6Lr4DeOAf+0GVgZLst2blgFqoqiH8dMqL9XgbC5dXT8lT+FDY1CvJCd
+cUJBQYPabjCbFLH1Q0r8NE19+QwCbZeS41wXFxmmHtB7pC+SC7237U035ATE6vSq
+guLNsRLd43VXQoG8PvEbVqvMQdYIht5KExvK4w81S0T2MugW5n6iYitjJsxlHk6Q
+8bN3wmty+MxGdBW+QbUqGhmS5Jle2HbyC2/hzuMyTOih5pntNF3DsH/gHL2J0rfu
+utZ89krUKDqs4WKfHebt8smVanh8PaKlTxr27z8aEp6YBBWkcRIhoBKJVTFKMLRJ
+XptBuyXNHVvHnz+C/dyGG/37W+jHwY8jE4n9jQFA7bVSqDjqeCxgNEMhQQARAQAB
+tB9Sb2dlciBMZWlnaCA8cmxlaWdoQGFwYWNoZS5vcmc+iQJUBBMBCAA+AhsDBQsJ
+CAcCBhUICQoLAgQWAgMBAh4BAheAFiEE8z0oHUcKtENnVhR8B7PIvECD6AAFAl2B
+RFgFCRJF2FYACgkQB7PIvECD6AC1hw//TnQuCzQgxaFzHQo/8InDGY4T+Uhb0O0X
+N0x/ojfMlkpb4BhXouzXiEuvSAfS0ySW/ZWJkWgfXwn+YDlB2AZxy8Znk0f5V6zN
+HZwqWnP1HWj5lCHCRHCfJGbzAxWlakSyUK70IF+arRFnkiQc64rpA0jcbAacEpPX
+qJU6keElHD7NmTDAyxhb1PzbcJ62Fz+789VE4J1rgUb9JUMYnAXdpLc2tkq4+Wyp
+E2sx35atYT4+9OuF7pYIH6RGCUxkhgYyihdc8msAlZ0WpB2BdLqEZae6iPNTMg3m
+syVhZK1Pj2J3BwM4cBOp13edduUamcWbhC7BT3tkPTroNncaoLh6id558ti8zdqC
+2v5Rg7TG7jAzrTB8vOCD+BvQFdDCKQZtGFOGEVxWJ0zaEO7iVnvHGM8uZ7kQVILD
+BWet3vJcDt4THJhEkbmN01+6sag9V9RCfRzDxEp/diJJAPYIV4R/MxJ/IN1coF44
+TLqspLG4UV6qKqDFvkrXX0QHXCF4pQvD2IWgB6DxUPkDWnF9jJQPm7tSeE86FsU1
+ZcTr7a4GHnLo0sy2dvwwrn7rDwiqJ8ar3KrBybLnZ2hiHU56Gsq1BernR6eyyDwi
+Q/LQINsQHnphWH1KE6vmi8crVrEYqHXDNWls9HSTTwmRr6uKSpTR/z2YuVxnCrUV
+PlcgvsDpZeq0IlJvZ2VyIExlaWdoIDxybGVpZ2hAY29kZWxpYnJlLm5ldD6JAlUE
+EwEIAD8CGwMCHgECF4AGCwkIBwMCBhUIAgkKCwQWAgMBFiEE8z0oHUcKtENnVhR8
+B7PIvECD6AAFAl2BRF0FCRJF2FYACgkQB7PIvECD6ADXwA//XTKg7/8mHqzUq9sA
+iUxQw4UQE25lUaAMMMIfH2tik6c1Y70hFC53xU+TDEb274WwfN3sUGGDX/EH2BQ4
+3goYG58PkVgxmlloUKpxyOeg/SjhL35UahK6ZOrgss6bdqdO/ZALziJpz8V6JPQa
+b9nqbAljuMDpza6a4233KqtURtFJXL/3tIbiSDjQ+2TVo0hkfuTpEORja3AcyYUS
+NNEup3zZPwlYIPODG3lVr5cEaIYLwP1TWYAisyzIOn2/VB9G/Fu7uF5u4DPyalEW
+Flbk5P+ko9Zk1gIp6A84mz1YegS++iZzSx+8RHaAS0tgxD3OUwIhTiW4UUku++bE
+uBTqn1lLgzxjDDGueQnvPBdMWKaJfS2TmyCkX2P4ZNPsaXhsn6TIuxi3ePWf1Y7J
+OJri+XG3NEKzGyGmolMF7jLmN1hoIaX3goHfkrIfec5RWidqq0k7iiCe17Ng93iI
+hDWe6Bc4N0r3gd4KZprLPxWN4wmhhPdV/g9MimLfw3iZk7KNPBUZLVTjD9zO3b4w
+WXapnFoZ2EK7u67ueH3/ffwYEnzwa0jlG6YzVERVJW5qPJjGOlanmljdmuTULFlc
+1jsB4Nbe0aBzoJipPxJHbxJe4nw/AAoJ8uihe6sXU3PfE6NavQxScJkJubV//MqI
+OB2uUaPX0dqbWC2f2DugERfCUpeIRgQQEQIABgUCTv3umAAKCRBVwVxpJb+4SGUm
+AKDy0gR2WZpSgrAiuEBT2o0wMzSxRwCgsFnoFNyTDV3BdUKL//kgF5N8o8GJAhwE
+EAECAAYFAk8fEFwACgkQ8WfkPIFDtoIrIhAAxnz09ZdA5DAcldsGtscSKmtzYwZz
+87EqYUqRMoxJBlbdpRiBnlrp3lOM82roue9WTTP2okDH8xnQnyfNet6SFxL0yPCq
+IDFmazKeW62fLN/bERqPuX9g5YRWL5MXisLtlB5M3p+E8SCn6yD5DhwvC09g1Tdb
+rxmbrP67nKYmUBhNCfvQczLhYC7DoCEBMz1lo2hNv+rwEJHvWqpH6NBP4oQaSsaz
+halFXwwmiG4yJdkRVdjaERWbQ4sw6MWhV4eiF26wiadU8ze/gPnsNAqK4PxLxSxf
+3jqkym2emSrmtJsjOaXbBnK8W49ynaPFOzKR9x+720rwRH9M+4iQ3+2ebO7YUs8h
+voXqocgbtsH+7tc+tc0s4jUm0oPrzf0F2rfcUweMb2eU+6EfdrHbOooEfLmpHnJp
+6mR96QY4Sa7rsvoYX7viKRhvgv7sdOJ+zBUILZEH7K1GDOWtmF8qLrjOJAmr/VrI
+0Wpmcw28l0W1gDJvSsJZhx70LoXcP2l6dtQj0AW4ulu5DpdRt2y+rnjylwcsOo0Y
+3TJBjiPB7ygOnbUT+P9EwFAmEajsa2c1MBYgp5eqi2bqhwpNLd/zBBNo6u0wj1KV
+zeu+m/NWAI5s4jukrI5ddEemnpmXNwJD61qMGRa2xM+3/+D0tGVTIZqt/TonZZy8
+4EKP0/w7ABpSKPmIRgQQEQIABgUCTx8QdQAKCRCIAQlKKLyz4y7ZAJ4jtVGP32cm
+B7nF3HpqoMQZcUSeVwCdHh9K1O0GSiPzwCwtZLluWQSFTC2JAhwEEAECAAYFAk8e
+6akACgkQI9DQutE9ekPhqg/+PdGxSK+FDz+hncVuVeSFtNcU1yYKJnAOq+ckDX60
+X87WQarm+s0WI6X6jGkrvxMHbbDBY3ghd37GU0cSbs95XOJ1xpHC4ujaXag6xp2h
+QdMjvbRmQ253wgVGQ1UfTWZMEM4+ksGKR4+GsGDZVgMqGVtTrFNIxoRSKH7kqmSH
+ZsGyZ8IsaEVgbRwjRDeWxz5QOsxaAI0fF6zkrEm5WCbNr+x+2laCXf+fIC3hr0hJ
+loWxa95rEDuuHiOFtiD6G1OCnK4SPxrbQnSS81l6U7dwLdL3YgDuhc8sqDP+dtBv
+sIZePkKzhMSadXAn/ff8lw36OxhoCpmcK46Fg/6n18fyrJqDa7DWoPRaU2budkL7
+Z0+S98+UAuYmE6dhnlNVZHoQMND5mAD6exFq7ToXfpEwwUZ6XT55QQhCLWc4D1+t
+B95M+cCel95xBpdYHAWQsxGuzbEcoPHmkP3Z/VH3+XI5hfNFBz66kFkuDqCGP6Bv
+WdrRidQuYHooAqFojvCr2ajz6tHUATlzjzg75nXrDNljLFXeUpS2Qfvm9i63blXV
+AttBpEtByhssQ12CYUE/86q7UC2KOts0gjqFCKigyDwG5Gl27uf0s8aA+YRcnCyn
+K00KkEDYXC+R4D8dUGdEtjaJAwXiWihGHu13xh56jnrJdwlJmsOxifavQ0Gx2aR9
+8VKJAhwEEwEIAAYFAk8fFY0ACgkQWHl5VzRCaE4MhA//Rms1fwJ2yd/NbTBmZ7T8
+5OL5M04Dw7UzuzqpOC/C4WMmKYQPW4UDuUuy6rTn4OU6uUXav85Fvgrsyy0Jkti0
+evBo5antM0FQxN2HIwjU1DLJyLLDNFGlwT4YIyqnxCJsCs/isKWqBr5uu/ysqil2
+OxQ+p8fMx46zZsx1QWgdrOxT7nK1UtrJ/YVtAFnZnZub1xIH6FXJCFghYuWCHlhT
+Apin1Df9mvjMdkM7vzu8QgfNyNHlL8SIgm9ikhw1LB9+TrsWaQNY2lATg5rMQA9d
+aij8uXiS6PxOU8U1xo0MRmLZmPdiH5B3vC4C4O1tDF1PKdpqUjujaf8wKVzP2SUV
+1TWqriXxp9Wpscvi1cbGUujJl9yOFgJKEN5KxalWsgl8mAHSmpc3eX6S8UJiiw8v
+gcC22FsIJuoucHh29loUwEyP2ai2+5V/0TPn/2byyhj/XUacgpQwMj51pu+Wd9OL
+FC3jBID3/hDQl0zp7tZztRlUswC0cKTRf4aUMmaEzoWANBWc/6rIkKLlUZ6UOPYK
+T22C+95pP0nRK/VZyN1xI7ikUFVFBHR5B49P2EXvUsFwrcK5hy5YRsT6xQkw69Gg
+KKechYn85YdH9kosRAVOTAQU/x+8CguodSkYvmp3nKdtyhXTheuIk+vrdFP+sqp/
+hrmqq0Aj9caCOdVdI4Epzq6JAhwEEAEKAAYFAk8fGFwACgkQLfv0Uj5UL9mCgg/+
+OZvaZL9XJ5z7EhJHsGhsTi+Sb1YpObomAbEpC+1jiQtjEnwbw6FnH0p9gxGzKnli
+wJ4nM5N4QL6JdZv3WL3rTu15XNWBk+YCrKY1MgtuSjvzJ0XJdrtQ+TnnU3Tnz1mu
+ZIIFE367NusF1ZbF3xOnqQcJKryrumIPhq1KFsn2Xi3kKzQBOA6SJ24Bb9vaw+MQ
+TmbvL2NBdvX5c/aDIWssAgZeiJG27/d7pGXTB1ajEDaHVJgGBxQPhwvykSf50Uoz
+lWuZ2Q73K6DOwbrg4fR0FZDPcsMFWlj15YIIrBmSFE2bXSkJDI+HVq7jikqNTGiv
+ek/YU0x8vCQnYqVFP9bjV7gO3S4rQxRR80cb/1QSYuIipaOENGvPSOw5/jLUmgHM
+763ckSY+OZT/wHdsZqEAVjcDdSrLkfKt/JdiyDmP6Dy9vZgfQ8Ia6R9Stwpz4y7M
+iOVYltTWNOCuGLvpJEunfGh8IJp1PGpfayrJq6sYKD9E8FcykLYwz0yjLb3kabwz
+nCfgpIoCWMTfcSGQu5EbbLeq1YMpMg2bWHO3S+sr3CjRjyDuo0bMa2d734x1TQuS
+uMhaYGar36MJjM+PU/2oGbfa3hSNnWtBH8S+DYbRV77yK9LlVo76XFVxRJXUCG3s
+eZQyOpDA2qXtlA24M97PMMmIrUM3YbreeERQJ0APtJCJAhwEEAECAAYFAk8fNCcA
+CgkQB7SEUnaNOCQxhg/9F2JiGM8SoEHdmIELgTp9+K05B4LSeowI5ueCwRstbi9N
+kCukvz3nT585pIqqQe5NDeuqBr3pUeCGSaqDwNiz/942wazBNV9XWUC7A0isgCOq
+ClJgfzjn4qjZ0BUKETY7gkflAHmvxxpDt8jJInfW5kWA4EFo++NzghEtnPTys55W
+IQ9OE9xbnW1pJpaO6cSZHfjgc82Z8ZpznYJMBqMfHxShflbkz+lAJLtVCRz3Wg8Q
+PCJ9mrN48PEnomTPMzBM4K21JtY7zqdy4MfYLl/PsqAsDJ4eOdzD4Hjnwd9sa5mK
+uZN7ukR8r3kmLrmPByTE/Q/sjMpNirdWfXd24wjdPAYNwTGxIS9ZPhvnT7MOh1cY
+cL49b0fp6aTIT/RrzakO4l0W1ZlaXN3TFQJNoOnea6xDHOkvdtM/j+ZjmsVBukuR
+djPpMKgoPM8mdNqCOJKh2v0c33r1ncGgNoEN8qAhuK+uNk/HzJ+aBb1kAJDIF22F
+ZXDdn4k0MDvLLKzAC6HNDC+xlix91HfjjHqSMadAikbmxt+zje8ROetgvsf53Iaf
+tUcjomA0gGpocIrQEl5xqj0MSU0Xsibt24e60SlU4GfN9/84SCzV9R4IJF/wX+fo
+/vH64s7fUp1r7zRJ0T7/QrJBZd0NlexkGr6wcZyUbJkQWjAXQhjVjh6kk8LgP86I
+RgQQEQgABgUCTyANTAAKCRCPY4+WGzBFzp9TAKCUTzk/UbwEphNFSjyerwTm74Q/
+AgCfVqq31zQUGRVWnFPjcmVwlRZZn1qJAhwEEAEIAAYFAk8gFNoACgkQ+gxBBzjJ
+2QN2lxAAh/d6GO3dbNEKYTbeqeAjTWebsEuQl1rAM0KeMJd45A1lkAkvdWxofxZf
+/RIqZZ/W7CGnMkIIg0beYR4SR+erf6llB+Rav8AFPURkivsFBjFYNa0yydLD1TNC
+vqRu/hdpSGU021Kk9jjMyozmrIWuuytq5sNQkRLxcYyQ7jYVeKf4p6MX5+X8u9zG
+8FRYA6V9sLh/CF4N186arLMwAIINrLNxbUKIFkiZGBtIT0b+VHQ5eLDA25UgT19R
+V6maw9oroNqOvaFQwI928VUDnEHnMN6gEVZQLBhI3KQQjRzHSe8KEzzEUIk8HN+2
+e552wbefquceifbmb1RwQMbPt0LxSTxJtoc59owWN7DkLO3j9uUkBbUgJvJnS3gl
+Mt/H30QVPZ3Kx6mKEc/dPJW7NgvMJtu1bgSNQnDZLTp/nze3nlM8gXbC4m1JvPOg
+T1YsXHZgYFqwT2scAXuhrnNgxOIKfdfpUpj76PorUY675qfN2YU6djSs247QK2CV
+skKOd562jEP4P+DMgNRtSMJXJyQwsFDK3XxeuuHdnWR7f9BJBoRD2cHiETXHaxgA
+ruWPfIboBojQb2Jiz1CkyiTY2/KCn0EAPie41IE3ylqaX6wnEPgqaLE4UOSBYZvU
+IP3XIXc/HkQpf/bVJ3T7SfTqxMRBnCJHqnZ67zO5xIh43nAixiyJAkkEEgEIADMF
+Ak8fPBcsGmh0dHA6Ly9uaWNrc2hhbmtzLmNvbS93ZWJvZnRydXN0I3BncC1wb2xp
+Y3kACgkQM9VSd0J1Q+C4uA//aE5+QqgyFO0JiaPIzUXc5m/bklMzqt0xcCWNHKgu
+d22tcQjitEj9YvF2TY2gRF37WhDDSxEJ30Hdk4dWm/1LNok5Fa9++1XscmJL3E3Y
+WXsHK5d/v5SRQeXNOiM4npezCvzUKiGXZw43hbIr/AlzMcd5TBjvPf0nzy0pyU6h
+3QarpomubvuCpmUkvypLFCij4qQZmicuG4q0KeEZRJ26+2vtxrf94ZJ7iI/tfRSf
+rPntVhMJW0JOIQB68mEKp2VGq3H7LieRY5fmXGrklGKQ5Xe6fO9+sRp+5izqtL9a
+XbjOXhhiAS1S8z/T8aHi2E3F5GrIkRdQP8ZVoAhPakmOc6fn2kiS0Zh85ib/znk1
+UeggQrUgAObgQxAnNHt6FMQG5yEc4vbhXrd6ZXynf1vYycF9jkFjWQhO0jbF/Cpp
+l6oa1Iwy0XGIoh5rZTbNXnARl/nlIvLRJAHqFZ7agKb+9qFCoWEWweIO3wcJBvz4
+xXw5qRvFAtu1LFe+dmKlwQPJVBheh6U/EwDXbMW3UNJG0TpQkIqCTRAWYZHxUnz7
+nb9GmrkvRO5/HL7O0DsLpqhG7YZWh4+JwzqLWq9aaM0rqzTxSZ7kezM/zzDd2giz
+wOkkYLrjlU2Yqim4NwWDTAgdOKR9Qrmxq+JsRpb/+y6xQC59xGnb1CSYPTwlUfRb
+viWJAhwEEwECAAYFAk8mzwgACgkQf3YGpEXcqA7xRA//SldDIsQCCpMmpHB4ACIN
+h8WlXpdklHqtqtDjinvFDOsZNYDbtBR85xfWx+Qt70A+jZXklVU4NYbG3jD3pIye
+b0uuz8D5+r+JmLMfdXcEmcC8Wx5hkwFSI3DCtxV8eXqZLh6QghVPTTUkSI5ZeiGF
+IKW3FbcTc6TUkRM2CqXjRlqLgJ3Igbws9PvlJWp4wfOyW+Qrj5vWnwH5fc5EosKY
+g8LxnWTMH0z79MAwZ+HYXVro3lm+3G6Zkid2Eeb1++NsEPOV01WCPfQVIlTg7KY6
+cYrC+D5a0/p0zTamgEs4V13Fja5/EAqXAsdNcFjLFLg3E6Mley7DVTDtu6h+jP6p
+4d1LoeSfgWQVQWoPqWTvAV26bX7xKeJV6Ccsv0i+0BvehVUTjKtVx8lfGwaSsE8f
+/iEvgyXHL+ULI+332StKil5AB0Eb/uPFDOFh6yAg/44jYEq9pey+Jmvhmcg6skKh
+U0m741E6W57z/sM+3KQ1+3yfDqn44htoOjoyqr8cxPQVbS76uiFR8B9rwqFzWAhP
+CQ9AtX2zu432nRTT1llbcZUnpJV48lz/+AWZRkTWyQPU57sZL+oG1LB7yNSG4+GA
+CFrBCVvWci4o6x5KjgEl2ZKQ7k9Lq99k/uLYUJtKdOolomNCS9/wpi9t6mCpt5j6
+ntSUXq55hE3wwk91Helm/COIRgQTEQIABgUCTybPsAAKCRArgHv7++dOmlGfAJ98
+3StiIR3UnkTrK5YEXADEb4KoSACg5E4Gm9StzNWOHRCCNpZikWEriUmJAhwEEwEC
+AAYFAk9+orgACgkQ1AZfXqfutgkBlA/+Pgza+wa+/wMUGdvmiIFIsW7KKG6DZsRF
+59i9uvlXeZDdGiLaBSosY8yM0NgDg//cVI7VrQyHRi7A5oayCMEYzcGE2w8q51XH
+/EyQkTXC8lgzhT3hgziTCnHbbdAL4zPtI9hQWiWVKx/4fo5jJ5p/+jKeOeAUEsob
++IbK8ZgcNtObBWbTz0pg4inWa+Pkm87QMTovEV0Po5MQMIbI1+0qrbIc1GozGSGl
+E16UXc8PPakzhc5/03pwm7G+XBI4+eVyHxL2GYKEDqvHewPjoJ3K/Qv0FrqYbpOu
+Sgtktkg8wRGP56BWbS3ABGdfZENgRY7716oiZJcFAyQ9mzOtUmRw3wzslwzx0QD+
+mC0foXrnXN6BAUW4NcNwWoM68RXbMUZRn2ngK/XKdrbjSNfTYuNHkyJ+mxOG4/7c
+NZAWHX6SeOtTA/YZP9GbekfH6lmjNYfO6kMgGhifNZddTKEtRUVg0wdUyaoT7s8s
+cxaAdjK7Q4mZqw5JSfL+SIxqOmv/WN+ptNfzfSaxuT5Vr7fKvyeZ26HhBr2gXD7k
+VOXJRo8RVnwyhctwZCQDNBjVQWjMafuDUrxmRKbnwMpz/DQtVZ7xEOUX2zlxkiD2
+Bpx0gOQchZuIfQukWbaRaI8l5wNE+Sozy8IsggH8ShfxVsJE1lgiL0uzQC3SISLi
+w5ERHqdDQ1OIRgQQEQgABgUCUDn3CQAKCRD0tLDMeX6/qyFqAKCAEx15iQTEFRw2
+IXur3dEPmrUZBgCeO1yYr2Gsd1hSMDSq9SivdmdNuV6JAhwEEAEIAAYFAlA59zAA
+CgkQY09L0eetVWibTw/7BnYLAo+OHkx+fUCIyWU7rbgNwg+6m81nFS1xJghIhQfN
+lU7Bxegc0RXikM0aQnpmtd5uC2UfGNfQnF7aXsQmFA/lKwUjJ/5DhWXNgOP3Q8Dr
+QLSSED7kFJzNDOS5lYvt6vba5h0gJBrNA1hIbEKOk22lORiQLGGu8moNn0ACeGGq
+DvtyD3vJNrjmTbxcVckq79HMhdyyW54vv0eUctpumvJlJKiIGSKSBKv93eeq643o
+PeoV2aCX53r4qgcvvhFuAxuq0DdX5S32ecBRzKEx4xDcg+WyMPIQnha2CETqcLRn
+/P8mhjh4yKRADvZUWjwQaykwtzx89ChoplbwJI1Uy8g5m0X8OgbLyG1kNbGgzjS5
+Gsdq/1FUdchv6n4MRZaRgRjc3WKIMJHdsAuh0cgS6+V7Ntw+TEPj4deDcZmg4R4U
+WPtrfAC3aus9TeU5qh7xpqZ6piHSRf55ezazTj61rtCaTKbZkTLWjQjaw62efeLM
+SVEKDLC0vlJ2OjlC09LeG1ASeLRbazZlP5KySocJxDKLSAwHEg0btJIUQ01p9hoH
+nRo9/tuKJPA9601m4/VKSBlaVe/4Piev9xU/mNL5Q4yyVPPEOvmAvlztEUIQwWPs
+ngHz9AigCrBiG/XH3Taws4lwZbta8WltDr1e6y1dXe2HfxVYadrSHiiCoS+KEjqJ
+AhwEEAEIAAYFAlA5/9wACgkQYDuDJmH5ylNRxRAAqHCoqyNRevQ/bqHg7U955jch
+EwSqeRcyBGICz6qN96LSjFIL2RS820DStzGjvFUSFuhYsBRGobOCL/92sA/owQd6
+4keBrxhShtTlmMOdp3SpjQrezHcxJsEwh1AJ2+MiEHEhuC/9j1z2nonW+DgVVrr1
+TfFyuf0ctIG9s2NPy7DeRsAf4SA8kO6QwQuFraOcL3u+kym7gvTKe1vAd5rmCvU/
+DuPWxAtpXiRilGXDNRGJbSHr8nSY5O9UFm5htCQrjlzUgxx/k9krOFbfslTh7YXj
+2yQ/xe8JKjc1ocq/RbH7nSqVz82EypPRV7w/xobt3szHkpneDCdQjc+H8RfRrezX
+EHmxemD5IoNTlqsIONCtzSj46P+rTdlNK8Gfr61cjOC6rTwZ6+Ao2SoYQRr7FvbW
+4Xd5XbVDVPUm3yD63I4wMrqS+j6OROIIRFxvLjsi0ROXczJtUfvjd4zNL6l7GxTX
+T+i+e/yDF7c6WY65ApHPL/klfqCW/L8ChvhE9HAy/BMIV1vl/3n3DPm8AxPLYBsx
+D4X6ICbzgqapWok5AD30Fkuyc4b62DQZJwLK/I+SSQlL10IMsAy927B6aLS1WRYP
+GNVbNZnXhVw8MF3sa9vpArZzI5C8q+0qdA5iknUjhjlrQhk7+NLCtIUnar9sd4SJ
+0m0rqaluLsJbTGgPEV2JAhwEEAEKAAYFAlBA2j0ACgkQU21XSJwgAdDIBRAAuYy6
+jUmdeDRwV06qFJqWAedKV1IEUuR1rDwf5nK9si5NBS4/2j2sZcrq9kMv/uN93CyU
+ZDdjRzz+NE9HXFqIjb8Jr6/wHPoh9bMZ6L+GauenxlK7UHtvDyzgTqf4mPSF87sp
+nXFrmeCyWU8knIn9nqXmhW357EktirQojGuOz/WOo9uEWFh9Cm8jKPCpc861p4ac
+8fRIEBpQhnsr87DWJc0+dxQcg4kx1cXmfx6cChy+sMzAhBa6Dyeu9/2Jb35oIscW
+yirIp7RmwVUFBEAMzPte1gedLvYGkpPJyzvRiRLX1ZJ4smR7g71ANgscvuBG5CHC
+VaMy+hn47T2bzJvlro2clLDZYqoDzVmED/gUYf5eP/dIgEdEPliVpSG9mdC7XNBK
+vHg9kv9wiU+Jkk6obTASE7mgldI7Rzp4yeULPYXDJVd1TMOTuLqbVVlf0cENcd09
+NXBFbBgSfER8huSlDFgPkiZ6yfdJV7cQ4GQuS3DV85caKiSx54T/d3AjBKkI30VF
+krmfqWu6jXjJwOG71y5b1wrFMGX9Fd5Mk72RX1JL9jSuDPU+CzPGP3ytVLzxB7hC
+cD6dcQUTL2kXn3dQ2Y/0mGSSusnWLCfCQ84CtAwv1A/RaP8vRkx+NXTp3Q8048q1
+jZInTqD3BZJRbtf1UEk4WWYoT25LdT38WYLJML6JAhwEEAEKAAYFAlA/6GAACgkQ
+1QplQSjT+IjjqQ/9H7vmK4qwlRsnce1d14UOgoRyPBiVFH9/Pz9B1StXwJcikSq6
+StYnJFW3KuZB8akXYK+wKa5Y7N4qm6/1bl3RjOJ7e/pyTvzgsFVuI3IBkxEg3GbK
+wxKmU51kBimOkY6OFO2g4v5WTMvBoMco4A2ldiTF2DVoSV4Di+Oj9J2qyuI/Zwpc
+lXWn9NsrlklHMm9nq/5WqAjY2uMmxY1Se9C7OUftiAx/7UEbavHvLMQ0/pysAoYm
+01TRrbX+uG8U9g0KwM9Hxm2U4guoXyI3S2qyUUR38QFpHcyJ+GqT+e4eRiskEaTK
+6OKbRGE1dHV1HtRYnvLJAeEgZmJmV35FLEL35XF8ptvjRLtzciBD6Z96VL2QI4Lb
+YQGgFFWxCcxu4JwPnJ5X1MFO0LMpsWKqf0d1DnlP8J6xz4Duu+3PBaVGOcYvRgpi
+inSpYe+WoKu4jqFxCBLU9PAoRqOSDEgkVDCOkeu243ZQ72bRKlJ+ZI/AvWQuwdUb
+yYXt5N8ZTaTqJQ9xhLse7uJ1BN7M5M5cWz3Yy9KvDD1GsQMEY240Wtm/74gnBAlC
+K6DBuNFIGhi9EnCC8b8/CI8/fNMZgeEFkRpvsRp48ievkCNvujRyuqGECkDrYacq
+ERP0P0I+ArQlMsOHN0qVF6EVKOQAmqyurL5sP2zlDjpS0G5astMfGfyzOUaJAhwE
+EAEIAAYFAlBbiFEACgkQxc5dwsVCzVm4ng/+KbUSk6fr7bxRR4hsRSZUQU6iwiDb
+kzmJyGWbvQn7YsDHZeCDOh1mgeICguAokbdfgl2tktcvbNN0xIFVzzWL0acUyqbM
+tJAMK9IpAK3DnOAFigeXmgUB8sjwddhLZUouD5o2frMSuoUgWp33I6/umQe/2Q+W
+UdkUYq5EEUwfmPVSxLKnJyh+hEEDfqvlfwBhK3AD9YMd4Uanw2xkCgeB3TTurKod
+yypy9RV5qCyra1yMqME5hs5DVRaoynVt3csb/fC2ldOQDzMq2hYMbr5Nvh2qwtEo
+Bs2m4SDsl2mqPPuSX5ZJ5KOceX/yujejCqAhDjCDNiD4OUf8Dfh2o8jjPQ7ICsbi
+tTnkZXdFnZHhtVLytjmKavRgTMJ7prOROz5LGhD/CPSjqQiw7d28vNdEXauf4ZB7
+GhLUVSGGHkvYc/RytFfRlcIxiA0CDwU3A3mthGuFwx096BUxFJiUCdOzh4pUAfMn
+6DbaXxy+d2XYR6JDIusrPXB4ic3kv41yyFxxhV+VAthZ9nAh4LsWn3JS6GjB7Y3g
+isnnfB2kUxWwH4oCbCjGlAV3xMKFoK1icCfzqR/BQqvsLYnzDpvk/7Pm3iWOtJmY
+7Ewu3bUBRjuSSxYozObftMZ6hYdeQK5HJdtqW/TMvceaH3+R2M4IQ7A4JuEK3Y/9
+IXv+6om1ejfPSmiJAVAEEwECADoFAlFPaV8zGmh0dHA6Ly9ueGZpZnRlZW4ubWUu
+dWsvZW4vYWJvdXQvZ3BnL3NpZ25pbmctcG9saWN5AAoJEGQVZ5VpqklGL6sH/A5J
+x2ilVGUJn7AF6VDAeTuC0pvVVmkb1aRvFjteBAedt8cd1DMUxGGDh1duoGfEC7cH
+IS9m2NWKhHhuXET1Bzq8wWb9gsFkViwUm6UmatnuJEZ22tWdqTkFQBR1neBjJQyC
+kMl7HGjShZPo42rN88NhREmAsnGZkA4XLOjYNroATjBjsp1OZtkfcW6/JMNe6P64
+og/YRYBtSOQ9bZdnshRL/6IePjvssO3OfGdjeEMptSY/zHWt9TXw8mdgul3C+fwi
+rIdr7mGJQgNbZ8vLH++V/4SBm+jUFPh0aAOear7eeuP3hs4aMgRjA46ejMrPdPSs
+y2yilG+QFW6SQId0/a25Ag0ETv3TAgEQALO7LRtq6VP8ffBhb4J5GaPyiab0iT6d
+2ckJjcqnzK7R4wWGlUg86cD6GW7Yqq+UQEkjPBEQQ7P2pa3WWSiPHbuwbkyCdWet
+knx9LN3uvfaO3x1JO8dW82f+zMz1fALkiK7ytJvWgf2PfDwjhjuxqBFiY+utcJqk
+PkFw5Y/cwdYeiLd1nTSDaFIOyYGT4uwqInSOZlfH4ilPB2fCfyckTQKsiKhpq9e6
+O1rzVHZWnjSQEysXXNAoLSXT4VjJ5NpSINp5rll0vTnAuAR41L8eHQl3gd19dg5o
+tUC+yIInklD/2lK4cKF0EPdrg2oOm+F/Y8U241wdBe1dlS/+1ljCnXwRhbBpTzbh
+7SPSePc6lunU0vW5fUWYnXQv+h6FpHrLuGLFSoXMzu7rfETgM206LNKPIAFChHYm
+KV/MX3j7u5iiuLy/JlSnjt82hv1JPQIWgZllqnPeP3ywuv9Rr/BglGt2WbUIpjGZ
+jr0m/eOShEQDJqdXffjPoK5di+BlCcXSLAOqVwS1xAP0yFphcx9E2dJGAw6IGxwh
+gvc9Qzg82oQ7IGKWu+vUicCch1ZyId7dNEd8iO9YIdk13EXlD64xfz0IvnBhHM3A
+KTWSbUgqKTQajYonmHjw24EwSMsfRvzOrseFtm6vp1WbxxG8NKLlsqhXLktEKUUm
+WDbj4DOAkhL/ABEBAAGJAiUEGAECAA8CGwwFAlLnwKQFCQesVJcACgkQB7PIvECD
+6AD1phAAp6P07TrINRc2cdSL8VX4aPzX/0FEwy1Tje2iEzuEB1nXhTYcGbzcMU7h
+/EvDMdOcB8HJBfBVZtf5tyOw+PZGqc20RWjCe21ak+4lxvNxrgnD/8FNPZLvESD1
+VWV6lxMfW9J5/fNo41VTa6kPRHJaxmhZsiA1NAmN2d9yMMOM1j7ZmaM2FwYC8StG
+QFnxidoCBQNprNLOIqaYsSJscMnOed8s9wpHnz717TRoNhcemdnv1uorF6rGIfBn
+Q5NBkZ3deEjVDz/hfsRJtvHFS8Nxc3cKwPolOIUni0/LaWSlnNHVm63F8v8IpFkz
+aQLXVTNEaLg+V7zhKQ6xw+sJ4upAb9Zc7HYPAYJ9iM7hPfvTXg9XjKvzxbAAkQFq
+bRXDLj3mnnmvJ8fTdlOVkhSKhbrJsmNco/XZGdza7MLH52Om/Ck3hk0WGfR6g+Uk
+UYTD31fQczZX7Fj0ivu6K62La5hBunCN5cdqrPYkIUVqd/0L0MQWg24IWbCNglIS
+yudk/C9k1Sc/VBcNtD/KDkEu8QxjxZSW//7svh0H3C4dI0QLgoAwMI/HGOhGwzCt
+eqdRCC/TmIJ1Wj463VE+Rd7CSLpimE9BgMid1I7AN+fVbk13jYvP/M+21T4fSnHR
+9hWjS2Ah32bNySBNBSAj+M+/1ooY4XgAoGk10Es7LZz8vEnSUJ+5Ag0ETv3vSwEQ
+AMk4qEW2SGTKreg7KGUCcX2als24bEL2vBKz4KSA/sWbZMrh68Bp3xviXy7ZPCxF
+G+zaSAsg3CXuZUMxmvugs/tUfWI5We+9owSXPOJFmNb6auyuiDlbV42WtuUpUMUN
+pCR2/SQR0KRHu0RT+lnA/Z5gcaegJZcpjjx0G2u98qxikkz/Q+A1eLNpbNozaseB
+CtG4GQWb1VycxV8b77VIa0o6pnQXJb+Wztr9r/e5ta7nqAw8Vcc2b+myMCj++5Vs
+MDcnY2bgPCd9jWaI3TtzZcAJdj9Iw2OmQxpwbqKo1RzyTDDedS1Og98XQEwZXYRy
+ZzhvB0U/8p+Tc35AJbXkGswGWpomDzaWZDIMH3bqQarLHt0PlcE2J1cMdJh/z2Ca
+0T5iq1jEPr7S+ru3rAtBkvgY9VVHabv41BTdw1odgT2ZIoXm0nD9VWYfyCnXaqHT
+yiQ4+ZFldL9O/FLnj+okxmJK8co9j2f7cDKC0mWT5DTFmkjPw6OT/in4fLWOz38F
+syM40zuYomdDHzIJZC4tnPpbcEnbuaVmD4+aG6UyIMqqP1dEiubTUImT8gdQs1nq
+wxe/W46MMrzqLbzSOQw9/AjLglsKD95fj+8aXUDUns6rYOsweXkZw1Cb6EB+qqze
+dq/S+e5PK7hkhTn/H8alGBm76PCsUz5LfH90rj9vfALdABEBAAGJBEQEGAECAA8C
+GwIFAlLnwMMFCQesOHMCKcFdIAQZAQIABgUCTv3vSwAKCRDiUkrFCp9cWaznEACG
+SN1EiNEEJ3zFMCrOyEXk51o2tnBIyEPTN2D+EPYnFR0ZWZT9Lkn2j05i7kevcF2m
+y8aIPnUaEQHOI691R2tt9wNSWZGXToNbEPZLZbn9knKJ5mnpQ47xMM3a3rTov0Th
+iRyjdNT/nOQiBCwF/vJ0V3jvzIugFRWtMGqDBBomfH5dT3nclETc3oLD27a9yHEq
+isc7BdcxgPBzhvwD1bTj09fg85aQPNCKb53K9i+4JWUrM+qRY/QSJd+Hg6NNAMpR
+WZPuHklWMKhPn+gH8D3abFJ4uy4ZtpRVg7O8VQWKAqNgRmYmhea5ITAW1Tmsyybh
+vRs4azcNFdmiMccfSnKXtPxML4pyTr95/AVZaamAZtPwyJ4fqP564YVAGf2Z2yM0
+pBbbsmK3T4GF0sEO33ushXOQuKvrBiEGdk3epIrqHsP0XpKyCbODAlDiz9NmQVyi
+jDMsBW84O7nbNbBoTmKc8ULOn60mBqoZmfzEO3TElIC0UYFRdUQBJ+G4j1uYjxoX
+4s/0UQlcVjOZfvEBQJdh4F0IBlEuHnzRAYJC57R3hn6ggOo2PX/16Ix+rI8sSPLX
+eSvS5qgXyi4p26xm3vm0OPcMTAz8HIA50F4aMTFCkKat6G0tVLHNlMo7dmA3nbBx
+P0XLqDyV1QvKU7ox7EG2Xp4biUZ6iGSDJ+VvhzbMsQkQB7PIvECD6AAq9Q/+OdnO
+j7wT3LLW4YxhxE6ip8Ok8vH2d44fQTACic+KAWl1aa5sHXLqrmAJVP1p9ofXPocz
+RrrSrWjwhgEm2YB+zubTIpmaTepKh/h5ozqMisQXZHl9E8ZaYl4WwGVQVTPdkqEY
+onwg5xtfUDOuCR0s90FiUR4xVFffP7ZB1PwziFATH6fCTy2JqPOIi6qu/isQgHHd
+sFyZrw3ycKxUYvzfrNJEYGiaeujh3yxFptlPNOed1lGIpEpuQT7Sdtq7r8j04ggm
+jC6pL7aqtq4b+yBM8nI3zN0wXAJOWR77l0PakeoQvVXG2Ac6r5wJY1Kod869o2HI
+YAypcUKWTqwvfS5wqflDg6bHFpDfQ+ytWqednUCZxfIBfGi5MkZOK9Jn+UGXUHw7
+YGYFku5S5FTh0ONeNu6vLKA84Bse+vY1m+Jk2i/dqr52OkI7ypZkZ8UCYePKCThP
+Wmm4kardCH/fR1z4wagHVjOKU5BS2fuCswEWf2+BWVSNAmNqXeAtKpgIfw8vw+ZC
+1Bqbwq+SX6QIRGZNWoK2rOlI+cfMGKnxcBm/ursXnaCTsZJ148Lw67Jl0u04vcBB
+z1vvbjnS1UihlGymCkI89B+o30V2yi8Hnfb4DPn9aLUrtirYVnNv//0/KLSyeR6K
+9oLdZjLs8W/yMk1lHIQ+JREswWQzc1SGIYVb2VC5AQ0EWBklIQEIAM/G04UA8dhT
+RHsoA5zITi6ZvR3btUSdgAMoiMG0UI/levVWmlsuqRiE+nXtr+JVGRJuRiaTGogh
+Arc2CGSBimE7CWunP/IFc0DkNIBLSQIQbuyfxy7GhL4uJQE/v6gdrkY5bg5CcQ4m
+BxQ/KaAudLli0lW/keS9t9BvBNxllB/MGnYxUW+vDcvMi8J6TTT2RHuvJK1v4Xo6
+3L9ReC5F4fP7V7FhBMGFQ9NC8K6iGGxX4D0tKYV/3D5bT0Sm1t53Gpt9APkzDX3H
+QB2a3zgrcrVb40yBAjDLAnhjBwBaQzNDXMYaonTy9nt2ha3X3d0/tqYfjo/X3MvW
+fKoclTixvHEAEQEAAYkDWwQYAQgAJgIbAhYhBPM9KB1HCrRDZ1YUfAezyLxAg+gA
+BQJdgUSZBQkJKoZ4ASnAXSAEGQEIAAYFAlgZJSEACgkQR7NR//YTiODfAAf9ERuk
+YFxvbRSdV54Cxqe6FVJDBlMVEheIBePj9GUPUWQAvQfosqRPtBt9hfYpSh6y0De0
+rN5O460UNLY2fmBNUl7TzMv+caKpmp3tjxK7B2t6+OI9uSiEVlTI5maLWvaKvAW5
+EIYXrdnpj/z7WKyaJnVowk3JZCrtNWIzYHh5EXHiiIrmOKa+ZGSmL9GWJsxp+hD6
+yFVKn5GP62wLID2AtT3rG6ExvyJLImQCetvwDsWpE4KS6q9avuvOcLWqCZKCajox
+ebR84P6FgAQQWFV5TsYuZgkt6KhrAz+kD9clqCcIt6lFHX4tganMQr+f5ZCwes4B
+tC6w6ZmCmjWfj4+wWwkQB7PIvECD6ACJ9A//VoJqhmtHSDyQyjkHCgHpZw0AzVCC
+KvX/SSgk7kMWLjbg5Qx7HeCHU9ieu8Xr8cz6Qju1knsIADeL9S6uqfG9J1HT+Xru
+nlaUuIHyyZdedgm1CA0HYZOvv1wbekbgY7J9tVHw/w40tE7+cTwsD2IgDHn2VUh3
+K5kZlSYBQkd8FTKpIOmM7kJZhA/d8KDvFLw8QQPZDwU5K/0mRNWcKajSMAe5x1gr
+5hQ26Yc7UFdkLLgPUJHGyoQOiK37/TEy1Fpz5rneiVn1kREqasUHtYz+xldco7pY
+LvhMuTCMhiYdpY+BkIqxlgfQFouuwhv5598z7g7lYURWE0x0/knhe/tEk3rXUbwW
+3sY3FSa5sp4ra0eJPP+25KU7IDaQPPldjU9gLhFdnaCI4O0boqGFSJgruvid3Cio
+GBIpoh4kOqePfS1WgCjxBr520k9wi+bB1d2/gPz99X+oz/wR395DhwFFiowDvaIM
+fhgeFUkL9Cpp3xZRr0K3xnHTq7Qv3ZJ2/4L8ZEA5Ie/mKm/CtMlA2dBCzNvE9vL9
+QJutdT++Snpc5hMT4NHfm7x4Kc+qMgkKBBf7MI0sCJ7+P+CmrsK8gOJ0NbgyKlyQ
+35nJcr6ddxoMMJqUVRNWhRsqHM/arsukawL3+AcrZz1h7cJ5BXoCnwfpIvN61HH7
+EkWRICdfQc/0aru5AQ0EWBklVQEIAL8LG7cFbEc7EOB+tU9CRKCj9OxutPertihb
+GOV2XX1R/Y6xY9FwxDmRRcp4573oEEA15fjlQpJoJnALUfwH9VNEwEAG4JQ3/Tcp
+bwIYNJ+Xzjt5kGohTd9ENWECjNv3LkZvS0usOjfC1DOaR7gqXBfB6eCdsnWph4ra
+/r5wdE8XJqbbe62Fn88r53f78b3FhNJUwOdWhlRSg5td+my7j9pMNEoJQ/VLr67v
+kTpQQKy3+GZsjPwz6K3s9oJ9zLO8b9UJl0GNWXmQ06E2LfHLsqfqFKuZO3Cwu+3S
+EnAqcgA7ftydV+wKZ8QpdF14jYnfkX+HAFCadxdbzcPS/6A1KM0AEQEAAYkCPAQY
+AQgAJgIbDBYhBPM9KB1HCrRDZ1YUfAezyLxAg+gABQJdgUSZBQkJKoZEAAoJEAez
+yLxAg+gA4asP/29sL9avyNuMPWIaBNI0J5tCnjehY8Xlkt0gcO3fgczDFclgsnBz
+Z7ChJyFVIHFo7GknsqF9g7N40z/EX1VxyHZ2s4/FH3WZNSn0g+X7HJtyVQ8JvEc7
+W6841lboJazJPsyZdKz9N21B3Lu7LfPQgkWPMmlhAVVz7s2AHrg+eBj4yyWwFGQ3
+d6FueRpAB0iRrVWX4TUg5xttOMRpCecjmMjHkSE7RA2uK7EfAr6m/o1k12ZisTI3
+6YgBp8sEAqOnPqsFZE5yKvNtowKn/OHw2AsAbVwWTqMmQtYBQGbNvTr+si7LAbcF
+w13ydR8SPGTZJ+Oh4osV7uhduJqTq6fFqCI/H4bOvvqwXzeR09Px/rcF538LqLQ2
++fjJb8HAMGU/ue5m0RzjcoKwWHC1Guuu8lo5wWjmNAnVRHXMVkO5gMds8EmNrZhP
+/xnHmAmhdaNRpwuEgEwPjfd5hFYwQqm2jFZtYd8MihfYf0JtQAQgyNMcXBTSnE5L
+WDii18axvlz6xPsyn/uoYXNBCLzAe7yjmy+QCe+Crsx9llC2VI4tGp+fOGde1J2E
+6HPvTjqT2nkefhb9exTVs/wTIDTMq26zUye810+LL0ErDLmHFvdOmnSNvpnX2A9O
+DXDRaMmdt7F05UmAyX0rXNARCUVuq1jxBVnazg7HW9UFfz8V0bMTD4qu
+=yJxr
+-----END PGP PUBLIC KEY BLOCK-----
diff --git a/NOTICE b/NOTICE
old mode 100755
new mode 100644
diff --git a/Tests/C-API/c-api.c b/Tests/C-API/c-api.c
index f38da9229..59fe44acf 100644
--- a/Tests/C-API/c-api.c
+++ b/Tests/C-API/c-api.c
@@ -25,13 +25,6 @@
 
 
 
-// This is here for memory leak testing.
-#if !defined(NDEBUG) && defined(_MSC_VER)
-#include <crtdbg.h>
-#endif
-
-
-
 void
 Usage()
 {
@@ -68,7 +61,7 @@ main(
             int     argc,
             char*   argv[])
 {
-#if !defined(NDEBUG) && defined(_MSC_VER)
+#if defined(XALAN_CRT_DEBUG)
     _CrtSetDbgFlag(_CrtSetDbgFlag(_CRTDBG_REPORT_FLAG) | _CRTDBG_LEAK_CHECK_DF);
 
     _CrtSetReportMode(_CRT_WARN, _CRTDBG_MODE_FILE);
diff --git a/Tests/Compare/compare.cpp b/Tests/Compare/compare.cpp
index 7c39856d7..ba80340ea 100644
--- a/Tests/Compare/compare.cpp
+++ b/Tests/Compare/compare.cpp
@@ -61,13 +61,6 @@ using std::endl;
 
 
 
-// This is here for memory leak testing.
-#if !defined(NDEBUG) && defined(_MSC_VER)
-#include <crtdbg.h>
-#endif
-
-
-
 using xalanc::XalanFileUtility;
 
 
@@ -219,7 +212,7 @@ main(
             int     argc,
             char*   argv[])
 {
-#if !defined(NDEBUG) && defined(_MSC_VER)
+#if defined(XALAN_CRT_DEBUG)
     _CrtSetDbgFlag(_CrtSetDbgFlag(_CRTDBG_REPORT_FLAG) | _CRTDBG_LEAK_CHECK_DF);
     _CrtSetReportMode(_CRT_WARN, _CRTDBG_MODE_FILE);
     _CrtSetReportFile(_CRT_WARN, _CRTDBG_FILE_STDERR);
diff --git a/Tests/Conf/conf.cpp b/Tests/Conf/conf.cpp
index c228a1db7..b328c0613 100644
--- a/Tests/Conf/conf.cpp
+++ b/Tests/Conf/conf.cpp
@@ -76,12 +76,6 @@ using namespace xalanc;
 
 
 
-// This is here for memory leak testing. 
-#if !defined(NDEBUG) && defined(_MSC_VER)
-#include <crtdbg.h>
-#endif
-
-
 
 using xercesc::MemoryManager;
 
@@ -619,7 +613,7 @@ main(
             int     argc,
             char*   argv[])
 {
-#if !defined(NDEBUG) && defined(_MSC_VER)
+#if defined(XALAN_CRT_DEBUG)
     _CrtSetDbgFlag(_CrtSetDbgFlag(_CRTDBG_REPORT_FLAG) | _CRTDBG_LEAK_CHECK_DF);
     _CrtSetReportMode(_CRT_WARN, _CRTDBG_MODE_FILE);
     _CrtSetReportFile(_CRT_WARN, _CRTDBG_FILE_STDERR);
diff --git a/Tests/Dom2Dom/dom2dom.cpp b/Tests/Dom2Dom/dom2dom.cpp
index f9b705505..5e1431031 100644
--- a/Tests/Dom2Dom/dom2dom.cpp
+++ b/Tests/Dom2Dom/dom2dom.cpp
@@ -32,11 +32,6 @@
 #include <unistd.h>
 #endif
 
-// This is here for memory leak testing. 
-#if !defined(NDEBUG) && defined(_MSC_VER)
-#include <crtdbg.h>
-#endif
-
 
 
 #include <xercesc/dom/DOMDocument.hpp>
@@ -387,7 +382,7 @@ main(
             int     argc,
             char*   argv[])
 {
-#if !defined(NDEBUG) && defined(_MSC_VER)
+#if defined(XALAN_CRT_DEBUG)
     _CrtSetDbgFlag(_CrtSetDbgFlag(_CRTDBG_REPORT_FLAG) | _CRTDBG_LEAK_CHECK_DF);
     _CrtSetReportMode(_CRT_WARN, _CRTDBG_MODE_FILE);
     _CrtSetReportFile(_CRT_WARN, _CRTDBG_FILE_STDERR);
diff --git a/Tests/Errortests/Errortests.cpp b/Tests/Errortests/Errortests.cpp
index e77eb23ec..5ad0f6932 100644
--- a/Tests/Errortests/Errortests.cpp
+++ b/Tests/Errortests/Errortests.cpp
@@ -56,12 +56,6 @@ using std::endl;
 #include <xalanc/Harness/XalanFileUtility.hpp>
 
 
-
-// This is here for memory leak testing. 
-#if !defined(NDEBUG) && defined(_MSC_VER)
-#include <crtdbg.h>
-#endif
-
 const char* const   excludeStylesheets[] =
 {
     0
@@ -310,7 +304,7 @@ main(
             int     argc,
             char*   argv[])
 {
-#if !defined(NDEBUG) && defined(_MSC_VER)
+#if defined(XALAN_CRT_DEBUG)
     _CrtSetDbgFlag(_CrtSetDbgFlag(_CRTDBG_REPORT_FLAG) | _CRTDBG_LEAK_CHECK_DF);
     _CrtSetReportMode(_CRT_WARN, _CRTDBG_MODE_FILE);
     _CrtSetReportFile(_CRT_WARN, _CRTDBG_FILE_STDERR);
diff --git a/Tests/Extensions/extensions.cpp b/Tests/Extensions/extensions.cpp
index f38cfe8c8..799e374d7 100644
--- a/Tests/Extensions/extensions.cpp
+++ b/Tests/Extensions/extensions.cpp
@@ -27,13 +27,6 @@
 
 
 
-// This is here for memory leak testing. 
-#if !defined(NDEBUG) && defined(_MSC_VER)
-#include <crtdbg.h>
-#endif
-
-
-
 #include <xercesc/util/PlatformUtils.hpp>
 
 
@@ -422,7 +415,7 @@ main(
             int     argc,
             char*   argv[])
 {
-#if !defined(NDEBUG) && defined(_MSC_VER)
+#if defined(XALAN_CRT_DEBUG)
     _CrtSetDbgFlag(_CrtSetDbgFlag(_CRTDBG_REPORT_FLAG) | _CRTDBG_LEAK_CHECK_DF);
     _CrtSetReportMode(_CRT_WARN, _CRTDBG_MODE_FILE);
     _CrtSetReportFile(_CRT_WARN, _CRTDBG_FILE_STDERR);
diff --git a/Tests/InputSource/inputsource.cpp b/Tests/InputSource/inputsource.cpp
index a047f5c2f..297944522 100644
--- a/Tests/InputSource/inputsource.cpp
+++ b/Tests/InputSource/inputsource.cpp
@@ -26,12 +26,6 @@
 
 
 
-// This is here for memory leak testing. 
-#if !defined(NDEBUG) && defined(_MSC_VER)
-#include <crtdbg.h>
-#endif
-
-
 
 // XERCES HEADERS...
 #include <xercesc/util/PlatformUtils.hpp>
@@ -416,7 +410,7 @@ main(
             int     argc,
             char*   argv[])
 {
-#if !defined(NDEBUG) && defined(_MSC_VER)
+#if defined(XALAN_CRT_DEBUG)
     _CrtSetDbgFlag(_CrtSetDbgFlag(_CRTDBG_REPORT_FLAG) | _CRTDBG_LEAK_CHECK_DF);
     _CrtSetReportMode(_CRT_WARN, _CRTDBG_MODE_FILE);
     _CrtSetReportFile(_CRT_WARN, _CRTDBG_FILE_STDERR);
diff --git a/Tests/Memory/Stressmem.cpp b/Tests/Memory/Stressmem.cpp
index f3205bd77..c78964c0b 100644
--- a/Tests/Memory/Stressmem.cpp
+++ b/Tests/Memory/Stressmem.cpp
@@ -25,13 +25,6 @@
 
 
 
-// This is here for memory leak testing. 
-#if !defined(NDEBUG) && defined(_MSC_VER)
-#include <crtdbg.h>
-#endif
-
-
-
 #include <xercesc/util/PlatformUtils.hpp>
 
 
@@ -230,7 +223,7 @@ main(
             int     argc,
             char*   argv[])
 {
-#if !defined(NDEBUG) && defined(_MSC_VER)
+#if defined(XALAN_CRT_DEBUG)
     _CrtSetDbgFlag(_CrtSetDbgFlag(_CRTDBG_REPORT_FLAG) | _CRTDBG_LEAK_CHECK_DF);
     _CrtSetReportMode(_CRT_WARN, _CRTDBG_MODE_FILE);
     _CrtSetReportFile(_CRT_WARN, _CRTDBG_FILE_STDERR);
diff --git a/Tests/Memory/TestBlock.cpp b/Tests/Memory/TestBlock.cpp
index 36a0184ec..7aa96eaf3 100644
--- a/Tests/Memory/TestBlock.cpp
+++ b/Tests/Memory/TestBlock.cpp
@@ -17,15 +17,7 @@
  */
 
 #if defined(_MSC_VER)
-
 #pragma warning(disable: 4127 4251 4511 4512 4514 4702 4710 4711 4786 4097; error: 4150 4172 4238 4239 4715)
-
-#if defined(_DEBUG)
-
-#include <crtdbg.h>
-
-#endif
-
 #endif
 
 
@@ -118,7 +110,7 @@ main(
 
             clock_t theStartClock = 0;
 
-#if defined(_MSC_VER) && defined(_DEBUG)
+#if defined(XALAN_CRT_DEBUG)
             // Send all reports to STDOUT
             _CrtSetReportMode(_CRT_WARN, _CRTDBG_MODE_FILE);
             _CrtSetReportFile(_CRT_WARN, _CRTDBG_FILE_STDOUT);
@@ -132,7 +124,7 @@ main(
             {
                 theStartClock = clock();
 
-#if defined(_MSC_VER) && defined(_DEBUG)
+#if defined(XALAN_CRT_DEBUG)
                 _CrtMemState    theStartState;
 
                 _CrtMemCheckpoint(&theStartState);
@@ -145,7 +137,7 @@ main(
                     theStringVector.push_back(m_allocator.create("Test string"));
                 }
 
-#if defined(_MSC_VER) && defined(_DEBUG)
+#if defined(XALAN_CRT_DEBUG)
                 _CrtMemState    theEndState;
 
                 _CrtMemCheckpoint(&theEndState);
@@ -180,7 +172,7 @@ main(
             theStringVector.clear();
 
             {
-#if defined(_MSC_VER) && defined(_DEBUG)
+#if defined(XALAN_CRT_DEBUG)
                 _CrtMemState    theStartState;
 
                 _CrtMemCheckpoint(&theStartState);
@@ -195,7 +187,7 @@ main(
                     theStringVector.push_back(m_allocator.create("Test string"));
                 }
 
-#if defined(_MSC_VER) && defined(_DEBUG)
+#if defined(XALAN_CRT_DEBUG)
                 _CrtMemState    theEndState;
 
                 _CrtMemCheckpoint(&theEndState);
@@ -237,7 +229,7 @@ main(
             {
                 theStartClock = clock();
 
-#if defined(_MSC_VER) && defined(_DEBUG)
+#if defined(XALAN_CRT_DEBUG)
                 _CrtMemState    theStartState;
 
                 _CrtMemCheckpoint(&theStartState);
@@ -248,7 +240,7 @@ main(
                     theStringVector.push_back(new string("Test string"));
                 }
 
-#if defined(_MSC_VER) && defined(_DEBUG)
+#if defined(XALAN_CRT_DEBUG)
                 _CrtMemState    theEndState;
 
                 _CrtMemCheckpoint(&theEndState);
diff --git a/Tests/Params/Params.cpp b/Tests/Params/Params.cpp
index 7cb288f0d..abb87fafe 100644
--- a/Tests/Params/Params.cpp
+++ b/Tests/Params/Params.cpp
@@ -24,11 +24,6 @@
 
 #include <iostream>
 
-// This is here for memory leak testing. 
-#if !defined(NDEBUG) && defined(_MSC_VER)
-#include <crtdbg.h>
-#endif
-
 
 
 #include <xercesc/util/PlatformUtils.hpp>
@@ -279,7 +274,7 @@ main(
             int     argc,
             char*   argv[])
 {
-#if !defined(NDEBUG) && defined(_MSC_VER)
+#if defined(XALAN_CRT_DEBUG)
     _CrtSetDbgFlag(_CrtSetDbgFlag(_CRTDBG_REPORT_FLAG) | _CRTDBG_LEAK_CHECK_DF);
     _CrtSetReportMode(_CRT_WARN, _CRTDBG_MODE_FILE);
     _CrtSetReportFile(_CRT_WARN, _CRTDBG_FILE_STDERR);
diff --git a/Tests/PerfT/perft.cpp b/Tests/PerfT/perft.cpp
index bbed2a472..d6a5a3db9 100644
--- a/Tests/PerfT/perft.cpp
+++ b/Tests/PerfT/perft.cpp
@@ -36,11 +36,6 @@
 #include <xalanc/Harness/XalanFileUtility.hpp>
 
 
-// This is here for memory leak testing.
-#if !defined(NDEBUG) && defined(_MSC_VER)
-#include <crtdbg.h>
-#endif
-
 
 
 using std::cerr;
@@ -387,7 +382,7 @@ main(
             int     argc,
             char*   argv[])
 {
-#if !defined(NDEBUG) && defined(_MSC_VER)
+#if defined(XALAN_CRT_DEBUG)
     _CrtSetDbgFlag(_CrtSetDbgFlag(_CRTDBG_REPORT_FLAG) | _CRTDBG_LEAK_CHECK_DF);
     _CrtSetReportMode(_CRT_WARN, _CRTDBG_MODE_FILE);
     _CrtSetReportFile(_CRT_WARN, _CRTDBG_FILE_STDERR);
diff --git a/Tests/Threads/ThreadTest.cpp b/Tests/Threads/ThreadTest.cpp
index f5b5a6787..cc53a7cea 100644
--- a/Tests/Threads/ThreadTest.cpp
+++ b/Tests/Threads/ThreadTest.cpp
@@ -86,12 +86,6 @@ using std::endl;
 using std::atoi;
 using std::strcmp;
 
-    
-    
-    // This is here for memory leak testing.
-#if !defined(NDEBUG) && defined(_MSC_VER)
-#include <crtdbg.h>
-#endif
 
 
 
@@ -661,7 +655,7 @@ main(
             int     argc,
             char*   argv[])
 {
-#if !defined(NDEBUG) && defined(_MSC_VER)
+#if defined(XALAN_CRT_DEBUG)
     _CrtSetDbgFlag(_CrtSetDbgFlag(_CRTDBG_REPORT_FLAG) | _CRTDBG_LEAK_CHECK_DF);
 
     _CrtSetReportMode(_CRT_WARN, _CRTDBG_MODE_FILE);
diff --git a/Tests/Transformer/TestXalanTransformer.cpp b/Tests/Transformer/TestXalanTransformer.cpp
index d9155fa44..618fb4b8e 100644
--- a/Tests/Transformer/TestXalanTransformer.cpp
+++ b/Tests/Transformer/TestXalanTransformer.cpp
@@ -23,11 +23,6 @@
 #define TEST_XALAN_CPP
 
 
-// This is here for memory leak testing.
-#if defined(_DEBUG)
-#include <crtdbg.h>
-#endif
-
 
 
 #include <xercesc/util/PlatformUtils.hpp>
@@ -306,7 +301,7 @@ main(
             int     /* argc */,
             char*   /* argv */ [])
 {
-#if !defined(NDEBUG) && defined(_MSC_VER)
+#if defined(XALAN_CRT_DEBUG)
     _CrtSetDbgFlag(_CrtSetDbgFlag(_CRTDBG_REPORT_FLAG) | _CRTDBG_LEAK_CHECK_DF);
     _CrtSetReportMode(_CRT_WARN, _CRTDBG_MODE_FILE);
     _CrtSetReportFile(_CRT_WARN, _CRTDBG_FILE_STDERR);
diff --git a/cmake/XalanDebug.cmake b/cmake/XalanDebug.cmake
new file mode 100644
index 000000000..b2bdbf63a
--- /dev/null
+++ b/cmake/XalanDebug.cmake
@@ -0,0 +1,36 @@
+# CMake build for xalan-c
+#
+# Written by Roger Leigh <rleigh@codelibre.net>
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Option for selection of shared or static libraries, exported as
+# cache variable
+
+set(xalan_debug_default OFF)
+if(CMAKE_BUILD_TYPE MATCHES Debug)
+  set(xalan_debug_default ON)
+endif()
+
+set(XALAN_DEBUG ${xalan_debug_default} CACHE BOOL "Build with additional debug checks")
+
+
+set(xalan_crt_debug_default OFF)
+if(xalan_debug_default AND MSVC)
+  set(xalan_crt_debug_default ON)
+endif()
+
+set(XALAN_CRT_DEBUG ${xalan_crt_debug_default} CACHE BOOL "Build with additional MSVC CRT debug checks (Windows only)")
diff --git a/samples/TransformToXercesDOM/TransformToXercesDOM.cpp b/samples/TransformToXercesDOM/TransformToXercesDOM.cpp
index c4082a610..89238b692 100644
--- a/samples/TransformToXercesDOM/TransformToXercesDOM.cpp
+++ b/samples/TransformToXercesDOM/TransformToXercesDOM.cpp
@@ -26,11 +26,6 @@
 
 
 
-#if !defined(NDEBUG) && defined(_MSC_VER)
-#include <crtdbg.h>
-#endif
-
-
 
 #include <xercesc/dom/DOMDocument.hpp>
 #include <xercesc/dom/DOMImplementation.hpp>
@@ -155,7 +150,7 @@ main(
             int     argc,
             char*   argv[])
 {
-#if !defined(NDEBUG) && defined(_MSC_VER)
+#if defined(XALAN_CRT_DEBUG)
     _CrtSetDbgFlag(_CrtSetDbgFlag(_CRTDBG_REPORT_FLAG) | _CRTDBG_LEAK_CHECK_DF);
 
     _CrtSetReportMode(_CRT_WARN, _CRTDBG_MODE_FILE);
diff --git a/samples/XPathWrapper/TestDriver.cpp b/samples/XPathWrapper/TestDriver.cpp
index efba36fd9..c4a4afb08 100644
--- a/samples/XPathWrapper/TestDriver.cpp
+++ b/samples/XPathWrapper/TestDriver.cpp
@@ -34,19 +34,13 @@
 
 
 
-// This is here for memory leak testing. 
-#if !defined(NDEBUG) && defined(_MSC_VER)
-#include <crtdbg.h>
-#endif
-
-
 
 int
 main(
             int     argc,
             char*   argv[])
 {
-#if !defined(NDEBUG) && defined(_MSC_VER)
+#if defined(XALAN_CRT_DEBUG)
     _CrtSetDbgFlag(_CrtSetDbgFlag(_CRTDBG_REPORT_FLAG) | _CRTDBG_LEAK_CHECK_DF);
     _CrtSetReportMode(_CRT_WARN, _CRTDBG_MODE_FILE);
     _CrtSetReportFile(_CRT_WARN, _CRTDBG_FILE_STDERR);
diff --git a/src/xalanc/CMakeLists.txt b/src/xalanc/CMakeLists.txt
index eee478e03..a5d9131c4 100644
--- a/src/xalanc/CMakeLists.txt
+++ b/src/xalanc/CMakeLists.txt
@@ -832,6 +832,7 @@ set(xalantransformer_sources
   XalanTransformer/XercesDOMWrapperParsedSource.cpp)
 
 set(xalantransformer_headers
+  XalanTransformer/XalanCAPI.h
   XalanTransformer/XalanCompiledStylesheetDefault.hpp
   XalanTransformer/XalanCompiledStylesheet.hpp
   XalanTransformer/XalanDefaultDocumentBuilder.hpp
diff --git a/src/xalanc/Harness/XalanFileUtility.cpp b/src/xalanc/Harness/XalanFileUtility.cpp
index 3a82ecc48..b0b166dcf 100644
--- a/src/xalanc/Harness/XalanFileUtility.cpp
+++ b/src/xalanc/Harness/XalanFileUtility.cpp
@@ -47,10 +47,6 @@
 #include <iostream>
 #include <strstream>
 
-#if !defined(NDEBUG) && defined(_MSC_VER)
-#include <crtdbg.h>
-#endif
-
 
 
 using std::fclose;
@@ -2119,7 +2115,7 @@ XalanFileUtility::analyzeResults(XalanTransformer& xalan, const XalanDOMString&
 
     if (result == 0)
     {
-#if defined(_MSC_VER)
+#if defined(XALAN_WINDOWS)
         CharVectorType theBuffer(getMemoryManager());
         TranscodeToLocalCodePage(theHTMLFile, theBuffer, true);
 
diff --git a/src/xalanc/Include/PlatformDefinitions.hpp.in b/src/xalanc/Include/PlatformDefinitions.hpp.in
index 57829a706..e460bf2d9 100644
--- a/src/xalanc/Include/PlatformDefinitions.hpp.in
+++ b/src/xalanc/Include/PlatformDefinitions.hpp.in
@@ -37,12 +37,6 @@
 #define XALAN_AUTO_PTR_REQUIRES_DEFINITION
 #endif
 
-#if defined(_DEBUG) && defined(NDEBUG)
-#error NDEBUG must not be defined when _DEBUG is defined.
-#elif !defined(_DEBUG) && !defined(NDEBUG)
-#error NDEBUG must be defined when _DEBUG is not defined.
-#endif
-
 #endif // _MSC_VER
 
 #if defined(__hpux)
@@ -59,7 +53,7 @@
 //  These defines provide the platform specific keywords that they need
 //  to do this.
 // ---------------------------------------------------------------------------
-#if defined(_MSC_VER)
+#if defined(XALAN_WINDOWS)
 #define XALAN_PLATFORM_EXPORT     __declspec(dllexport)
 #define XALAN_PLATFORM_IMPORT     __declspec(dllimport)
 #define XALAN_PLATFORM_EXPORT_FUNCTION(T) XALAN_PLATFORM_EXPORT T
@@ -97,6 +91,13 @@
 
 #cmakedefine XALAN_NON_ASCII_PLATFORM 1
 
+#cmakedefine XALAN_DEBUG 1
+#cmakedefine XALAN_CRT_DEBUG 1
+
+#if defined(XALAN_CRT_DEBUG)
+#include <crtdbg.h>
+#endif
+
 #include <xalanc/Include/XalanVersion.hpp>
 
 #if defined(__cplusplus)
diff --git a/src/xalanc/Include/XalanList.hpp b/src/xalanc/Include/XalanList.hpp
index cf72f4626..f31950d28 100644
--- a/src/xalanc/Include/XalanList.hpp
+++ b/src/xalanc/Include/XalanList.hpp
@@ -359,14 +359,11 @@ public:
     void 
     splice(
             iterator    pos,
-#if defined(NDEBUG)
-            ThisType&   /* list */,
-#else
             ThisType&   list,
-#endif
             iterator    toInsert)
     {
         assert(m_memoryManager == list.m_memoryManager);
+        (void) list; // unused if assert is disabled
 
         if (pos != toInsert)
         {
@@ -387,15 +384,12 @@ public:
     void
     splice(
             iterator    pos,
-#if defined(NDEBUG)
-            ThisType&   /* list */,
-#else
             ThisType&   list,
-#endif
             iterator    toInsertFirst,
             iterator    toInsertLast)
     {
         assert(m_memoryManager == list.m_memoryManager);
+        (void) list; // unused if assert is disabled
 
         if (toInsertFirst != toInsertLast)
         {
diff --git a/src/xalanc/Include/XalanVector.hpp b/src/xalanc/Include/XalanVector.hpp
index b8ed46d4c..7571dfff5 100644
--- a/src/xalanc/Include/XalanVector.hpp
+++ b/src/xalanc/Include/XalanVector.hpp
@@ -868,12 +868,6 @@ public:
 
 private:
 
-#if defined(NDEBUG)
-    void
-    invariants() const
-    {
-    }
-#else
     void
     invariants() const
     {
@@ -882,7 +876,6 @@ private:
             (m_data == 0 && m_allocation == 0) ||
             (m_data != 0 && m_allocation != 0));
     }
-#endif
 
     size_type
     local_distance(
diff --git a/src/xalanc/PlatformSupport/ArenaBlock.hpp b/src/xalanc/PlatformSupport/ArenaBlock.hpp
index eb464a4ec..676080715 100644
--- a/src/xalanc/PlatformSupport/ArenaBlock.hpp
+++ b/src/xalanc/PlatformSupport/ArenaBlock.hpp
@@ -111,14 +111,11 @@ public:
      * @param theBlock the address that was returned by allocateBlock()
      */
     void
-#if defined (NDEBUG)
-    commitAllocation(ObjectType*    /* theBlock */)
-#else
     commitAllocation(ObjectType*    theBlock)
-#endif
     {
         assert(theBlock == this->m_objectBlock + this->m_objectCount);
         assert(this->m_objectCount < this->m_blockSize);
+        (void) theBlock; // unused if assert is disabled
 
         ++this->m_objectCount;
     }
diff --git a/src/xalanc/PlatformSupport/DirectoryEnumerator.hpp b/src/xalanc/PlatformSupport/DirectoryEnumerator.hpp
index 2384c7072..ffd5e2737 100644
--- a/src/xalanc/PlatformSupport/DirectoryEnumerator.hpp
+++ b/src/xalanc/PlatformSupport/DirectoryEnumerator.hpp
@@ -26,7 +26,7 @@
 
 
 #include <cstring>
-#if defined(_MSC_VER)
+#if defined(XALAN_WINDOWS)
 #include <io.h>
 #else
 #include <dirent.h>
@@ -59,7 +59,7 @@ using xercesc::MemoryManager;
 
 
 
-#if defined(_MSC_VER)
+#if defined(XALAN_WINDOWS)
 
 struct FindFileStruct : public _wfinddata_t
 {
@@ -240,7 +240,7 @@ EnumerateDirectory(
             bool                        fIncludeSelfAndParent = false)
 #endif
 {
-#if defined(_MSC_VER)
+#if defined(XALAN_WINDOWS)
     FindFileStruct      theFindData;
     
     #ifdef _WIN64
@@ -249,13 +249,17 @@ EnumerateDirectory(
         typedef long    theHandleType;
     #endif
 
+#if defined(_MSC_VER)
 #pragma warning(push)
 #pragma warning(disable: 4244)
+#endif
     theHandleType   theSearchHandle =
         _wfindfirst(
             reinterpret_cast<wchar_t*>(const_cast<XalanDOMChar*>(theConversionFunction(theFullSearchSpec))),
             &theFindData);
+#if defined(_MSC_VER)
 #pragma warning(pop)
+#endif
 
     if (theSearchHandle != -1)
     {
diff --git a/src/xalanc/PlatformSupport/DoubleSupport.hpp b/src/xalanc/PlatformSupport/DoubleSupport.hpp
index a14235451..e13393b5e 100644
--- a/src/xalanc/PlatformSupport/DoubleSupport.hpp
+++ b/src/xalanc/PlatformSupport/DoubleSupport.hpp
@@ -25,7 +25,7 @@
 
 
 
-#if defined(_MSC_VER)
+#if defined(XALAN_WINDOWS)
 #include <float.h>
 #endif
 #include <cmath>
diff --git a/src/xalanc/PlatformSupport/XalanDOMStringHashTable.cpp b/src/xalanc/PlatformSupport/XalanDOMStringHashTable.cpp
index ee3381387..559ffd1ed 100644
--- a/src/xalanc/PlatformSupport/XalanDOMStringHashTable.cpp
+++ b/src/xalanc/PlatformSupport/XalanDOMStringHashTable.cpp
@@ -57,7 +57,7 @@ XalanDOMStringHashTable::clear()
         m_buckets[i].clear();
     }
 
-#if !defined(NDEBUG)
+#if defined(XALAN_DEBUG)
     m_collisions = 0;
 #endif
     m_count = 0;
@@ -181,7 +181,7 @@ XalanDOMStringHashTable::insert(const XalanDOMString&   theString)
 
     BucketType& theBucket = m_buckets[theBucketIndex];
 
-#if !defined(NDEBUG)
+#if defined(XALAN_DEBUG)
     if (theBucket.empty() == false)
     {
         ++m_collisions;
@@ -207,7 +207,7 @@ XalanDOMStringHashTable::insert(
 
     BucketType&     theBucket = m_buckets[theBucketIndex];
 
-#if !defined(NDEBUG)
+#if defined(XALAN_DEBUG)
     if (theBucket.empty() == false)
     {
         ++m_collisions;
diff --git a/src/xalanc/PlatformSupport/XalanFileOutputStream.hpp b/src/xalanc/PlatformSupport/XalanFileOutputStream.hpp
index 27ab4d7f8..3376fa329 100644
--- a/src/xalanc/PlatformSupport/XalanFileOutputStream.hpp
+++ b/src/xalanc/PlatformSupport/XalanFileOutputStream.hpp
@@ -40,8 +40,9 @@
 
 namespace XALAN_CPP_NAMESPACE {
 
-
+#if !defined(XALAN_WINDOWS)
 using std::FILE;
+#endif
 
 class XALAN_PLATFORMSUPPORT_EXPORT XalanFileOutputStream : public XalanOutputStream
 {
diff --git a/src/xalanc/PlatformSupport/XalanICUMessageLoader.cpp b/src/xalanc/PlatformSupport/XalanICUMessageLoader.cpp
index 6dd2e67ec..aa18d1b1b 100644
--- a/src/xalanc/PlatformSupport/XalanICUMessageLoader.cpp
+++ b/src/xalanc/PlatformSupport/XalanICUMessageLoader.cpp
@@ -49,30 +49,9 @@
 
 #if defined(XALAN_WINDOWS)
 
-#if _MSC_VER >= 1400
-
 #define ENTRY_POINT_SUFFIX _dat
 #define PACKAGE_NAME_SUFFIX
 
-#else
-#ifdef  NDEBUG_SYMBOLS // release with symbols
-
-#define ENTRY_POINT_SUFFIX S_dat
-#define PACKAGE_NAME_SUFFIX S
-
-#elif NDEBUG // release
-
-#define ENTRY_POINT_SUFFIX _dat
-#define PACKAGE_NAME_SUFFIX
-
-#elif _DEBUG // debug only 
-
-#define ENTRY_POINT_SUFFIX D_dat
-#define PACKAGE_NAME_SUFFIX D
-
-#endif // NDEBUG_SYMBOLS
-#endif // _MSC_VER >= 1400
-
 #define PACKAGE_NAME INVK_CAT3_RAW_NUMERIC(XALAN_MESSAGES_NAME_W_UNDERSCORE,\
                                     INVK_CAT2_RAW_NUMERIC_SEP_UNDERSCORE(XALAN_VERSION_MAJOR,XALAN_VERSION_MINOR),\
                                     PACKAGE_NAME_SUFFIX)
diff --git a/src/xalanc/PlatformSupport/XalanReferenceCountedObject.cpp b/src/xalanc/PlatformSupport/XalanReferenceCountedObject.cpp
index c020f1ad7..7f6d09bde 100644
--- a/src/xalanc/PlatformSupport/XalanReferenceCountedObject.cpp
+++ b/src/xalanc/PlatformSupport/XalanReferenceCountedObject.cpp
@@ -22,10 +22,7 @@
 
 
 #include <cassert>
-
-#if !defined(NDEBUG)
 #include <climits>
-#endif
 
 
 
diff --git a/src/xalanc/TestXPath/TestXPath.cpp b/src/xalanc/TestXPath/TestXPath.cpp
index 721c3b39c..5c7f26280 100644
--- a/src/xalanc/TestXPath/TestXPath.cpp
+++ b/src/xalanc/TestXPath/TestXPath.cpp
@@ -31,12 +31,6 @@
 
 
 
-#if !defined(NDEBUG) && defined(_MSC_VER)
-#include <crtdbg.h>
-#endif
-
-
-
 #include <xercesc/framework/URLInputSource.hpp>
 #include <xercesc/util/PlatformUtils.hpp>
 
@@ -1339,7 +1333,7 @@ main(
 {
     using std::cout;
 
-#if !defined(NDEBUG) && defined(_MSC_VER)
+#if defined(XALAN_CRT_DEBUG)
     _CrtSetDbgFlag(_CrtSetDbgFlag(_CRTDBG_REPORT_FLAG) | _CRTDBG_LEAK_CHECK_DF);
 
    _CrtSetReportMode(_CRT_WARN, _CRTDBG_MODE_FILE);
diff --git a/src/xalanc/TestXSLT/process.cpp b/src/xalanc/TestXSLT/process.cpp
index 026afb6a0..da9d1322f 100644
--- a/src/xalanc/TestXSLT/process.cpp
+++ b/src/xalanc/TestXSLT/process.cpp
@@ -28,12 +28,6 @@
 
 
 
-#if !defined(NDEBUG) && defined(_MSC_VER)
-#include <crtdbg.h>
-#endif
-
-
-
 #include <xercesc/util/PlatformUtils.hpp>
 #include <xercesc/sax/SAXParseException.hpp>
 
@@ -1092,7 +1086,7 @@ main(
             int     argc,
             char*   argv[])
 {
-#if !defined(NDEBUG) && defined(_MSC_VER)
+#if defined(XALAN_CRT_DEBUG)
     _CrtSetDbgFlag(_CrtSetDbgFlag(_CRTDBG_REPORT_FLAG) | _CRTDBG_LEAK_CHECK_DF);
 
     _CrtSetReportMode(_CRT_WARN, _CRTDBG_MODE_FILE);
diff --git a/src/xalanc/Utils/MsgCreator/MsgCreator.cpp b/src/xalanc/Utils/MsgCreator/MsgCreator.cpp
index c742389b0..0f84010eb 100644
--- a/src/xalanc/Utils/MsgCreator/MsgCreator.cpp
+++ b/src/xalanc/Utils/MsgCreator/MsgCreator.cpp
@@ -29,9 +29,7 @@
 #include <xercesc/util/TransService.hpp>
 
 #include <cassert>
-#if defined(_DEBUG) && defined(_MSC_VER)
-#include <crtdbg.h>
-#endif
+
 
 
 using std::cerr;
@@ -327,7 +325,7 @@ buildIndexFileName(
 int main(int argC, char* argV[])
 {
     
-#if defined(_DEBUG) && defined(_MSC_VER)
+#if defined(XALAN_CRT_DEBUG)
     _CrtSetDbgFlag(_CrtSetDbgFlag(_CRTDBG_REPORT_FLAG) | _CRTDBG_LEAK_CHECK_DF);
     
     _CrtSetReportMode(_CRT_WARN, _CRTDBG_MODE_FILE);
diff --git a/src/xalanc/XMLSupport/XalanUTF8Writer.hpp b/src/xalanc/XMLSupport/XalanUTF8Writer.hpp
index 15ece478f..65888fead 100644
--- a/src/xalanc/XMLSupport/XalanUTF8Writer.hpp
+++ b/src/xalanc/XMLSupport/XalanUTF8Writer.hpp
@@ -201,7 +201,7 @@ public:
             const value_type*   theChars,
             size_type           theLength)
     {
-    #if defined(NDEBUG)
+    #if !defined(XALAN_DEBUG)
         if (theLength > sizeof(m_buffer))
         {
             flushBuffer();
diff --git a/src/xalanc/XPath/MutableNodeRefList.cpp b/src/xalanc/XPath/MutableNodeRefList.cpp
index 9d8c1ba1d..f7e352780 100644
--- a/src/xalanc/XPath/MutableNodeRefList.cpp
+++ b/src/xalanc/XPath/MutableNodeRefList.cpp
@@ -704,7 +704,9 @@ MutableNodeRefList::clearNulls()
         m_order = eUnknownOrder;
     }
 
+#if defined(XALAN_DEBUG)
     assert(checkForDuplicates(getMemoryManager()) == false);
+#endif
 }
 
 
diff --git a/src/xalanc/XPath/NodeRefList.cpp b/src/xalanc/XPath/NodeRefList.cpp
index a71cb6654..a35a2be3a 100644
--- a/src/xalanc/XPath/NodeRefList.cpp
+++ b/src/xalanc/XPath/NodeRefList.cpp
@@ -23,9 +23,7 @@
 #include <cassert>
 #include <algorithm>
 
-#if !defined(NDEBUG)
 #include <xalanc/Include/XalanSet.hpp>
-#endif
 
 
 
@@ -151,7 +149,7 @@ NodeRefList::indexOf(const XalanNode*   theNode) const
 
 
 
-#if !defined(NDEBUG)
+#if defined(XALAN_DEBUG)
 bool
 NodeRefList::checkForDuplicates(MemoryManager& theManager) const
 {
diff --git a/src/xalanc/XPath/NodeRefList.hpp b/src/xalanc/XPath/NodeRefList.hpp
index 058e1827a..0f6fd39cf 100644
--- a/src/xalanc/XPath/NodeRefList.hpp
+++ b/src/xalanc/XPath/NodeRefList.hpp
@@ -99,7 +99,7 @@ public:
     virtual size_type
     indexOf(const XalanNode*    theNode) const;
 
-#if !defined(NDEBUG)
+#if defined(XALAN_DEBUG)
     bool
     checkForDuplicates(MemoryManager& theManager) const;
 #endif
diff --git a/src/xalanc/XPath/XPathExpression.cpp b/src/xalanc/XPath/XPathExpression.cpp
index 6d05b2350..ab5eb6638 100644
--- a/src/xalanc/XPath/XPathExpression.cpp
+++ b/src/xalanc/XPath/XPathExpression.cpp
@@ -629,11 +629,7 @@ XPathExpression::insertOpCode(
 void
 XPathExpression::updateShiftedOpCodeLength(
             OpCodeMapValueType  theOpCode,
-#if defined(NDEBUG)
-            OpCodeMapSizeType   /* theOriginalIndex */,
-#else
             OpCodeMapSizeType   theOriginalIndex,
-#endif
             OpCodeMapSizeType   theNewIndex)
 {
     // There must be some other expressions in
@@ -641,6 +637,7 @@ XPathExpression::updateShiftedOpCodeLength(
     assert(opCodeMapSize() > theNewIndex + 1);
 
     assert(theNewIndex > theOriginalIndex);
+    (void) theOriginalIndex; // unused if assert is disabled
 
     const OpCodeMapValueType    theOpCodeLength = getOpCodeLength(theOpCode);
 
diff --git a/src/xalanc/XPath/XalanDocumentFragmentNodeRefListBaseProxy.cpp b/src/xalanc/XPath/XalanDocumentFragmentNodeRefListBaseProxy.cpp
index da552131c..6bfaf2828 100644
--- a/src/xalanc/XPath/XalanDocumentFragmentNodeRefListBaseProxy.cpp
+++ b/src/xalanc/XPath/XalanDocumentFragmentNodeRefListBaseProxy.cpp
@@ -56,13 +56,10 @@ XalanDocumentFragmentNodeRefListBaseProxy::~XalanDocumentFragmentNodeRefListBase
 
 
 XalanNode*
-#if defined(NDEBUG)
-XalanDocumentFragmentNodeRefListBaseProxy::item(size_type) const
-#else
 XalanDocumentFragmentNodeRefListBaseProxy::item(size_type   index) const
-#endif
 {
     assert(index == 0);
+    (void) index; // unused if assert is disabled
 
     return const_cast<XalanDocumentFragment*>(&m_value);
 }
diff --git a/src/xalanc/XPathCAPI/XPathCAPI.h b/src/xalanc/XPathCAPI/XPathCAPI.h
index b754c0054..af95148e5 100644
--- a/src/xalanc/XPathCAPI/XPathCAPI.h
+++ b/src/xalanc/XPathCAPI/XPathCAPI.h
@@ -20,7 +20,7 @@
 
 
 
-#if defined(_MSC_VER)
+#if defined(XALAN_WINDOWS)
 
 #if defined(XALAN_BUILD_DLL)
 
diff --git a/src/xalanc/XSLT/FunctionDocument.cpp b/src/xalanc/XSLT/FunctionDocument.cpp
index bd02cffe7..701f982f1 100644
--- a/src/xalanc/XSLT/FunctionDocument.cpp
+++ b/src/xalanc/XSLT/FunctionDocument.cpp
@@ -573,7 +573,9 @@ FunctionDocument::doExecute(
         }
     }
 
+#if defined(XALAN_DEBUG)
     assert(mnl->checkForDuplicates(executionContext.getMemoryManager()) == false);
+#endif
 
     mnl->setDocumentOrder();
 
diff --git a/src/xalanc/XSLT/VariablesStack.cpp b/src/xalanc/XSLT/VariablesStack.cpp
index 222658ec4..cd6fa7046 100644
--- a/src/xalanc/XSLT/VariablesStack.cpp
+++ b/src/xalanc/XSLT/VariablesStack.cpp
@@ -523,7 +523,7 @@ VariablesStack::pushElementFrame(const ElemTemplateElement* elem)
 {
     push(StackEntry(elem));
 
-#if !defined(NDEBUG)
+#if defined(XALAN_DEBUG)
     m_elementFrameStack.push_back(elem);
 #endif
 }
@@ -575,7 +575,7 @@ VariablesStack::popElementFrame()
         }
         else if (theEntry.getType() == StackEntry::eElementFrameMarker)
         {
-#if !defined(NDEBUG)
+#if defined(XALAN_DEBUG)
             const ElemTemplateElement* const    theElement =
                 theEntry.getElement();
 
diff --git a/src/xalanc/XalanDOM/XalanDOMString.hpp b/src/xalanc/XalanDOM/XalanDOMString.hpp
index 270b9317f..f1b18b804 100644
--- a/src/xalanc/XalanDOM/XalanDOMString.hpp
+++ b/src/xalanc/XalanDOM/XalanDOMString.hpp
@@ -772,10 +772,8 @@ protected:
     void
     invariants() const
     {
-#if !defined(NDEBUG)
         assert((m_data.empty() == true && m_size == 0) || m_size == m_data.size() - 1);
         assert(m_data.empty() == true || m_data.back() == 0);
-#endif
     }
 
     /*
diff --git a/src/xalanc/XalanDOM/XalanNode.hpp b/src/xalanc/XalanDOM/XalanNode.hpp
index befb40b71..61552b27e 100644
--- a/src/xalanc/XalanDOM/XalanNode.hpp
+++ b/src/xalanc/XalanDOM/XalanNode.hpp
@@ -187,12 +187,6 @@ protected:
 
     bool
     operator==(const XalanNode&     theRHS) const;
-
-private:
-
-#if !defined(NDEBUG)
-    static size_t   s_instanceCount;
-#endif
 };
 
 
diff --git a/src/xalanc/XalanExe/XalanExe.cpp b/src/xalanc/XalanExe/XalanExe.cpp
index 90823ce13..00f1c0487 100644
--- a/src/xalanc/XalanExe/XalanExe.cpp
+++ b/src/xalanc/XalanExe/XalanExe.cpp
@@ -26,12 +26,6 @@
 
 
 
-#if !defined(NDEBUG) && defined(_MSC_VER)
-#include <crtdbg.h>
-#endif
-
-
-
 #include <xercesc/util/PlatformUtils.hpp>
 #include <xercesc/util/OutOfMemoryException.hpp>
 
@@ -50,7 +44,7 @@
 
 
 
-#if defined(_MSC_VER)
+#if defined(XALAN_WINDOWS)
 #define XALAN_USE_WINDOWS_TIMING
 #endif
 
@@ -896,7 +890,7 @@ xsltMain(
 
     using xercesc::XMLPlatformUtils;
 
-#if defined(XALAN_WINDOWS) && defined(NDEBUG)
+#if defined(XALAN_WINDOWS) && !defined(XALAN_DEBUG)
     WindowsMemoryManager  theMemoryManager;
 
     // Call the static initializer for Xerces...
@@ -969,7 +963,7 @@ main(
             int     argc,
             char*   argv[])
  {
-#if defined(_DEBUG) && defined(_MSC_VER)
+#if defined(XALAN_CRT_DEBUG)
     _CrtSetDbgFlag(
         _CrtSetDbgFlag(_CRTDBG_REPORT_FLAG) | _CRTDBG_LEAK_CHECK_DF);
 
